crates/typst/src/layout/grid/layout.rs

```diff
@@ -216,38 +226,123 @@ impl CellGrid {
         inset: Sides<Rel<Length>>,
         engine: &mut Engine,
         styles: StyleChain,
+        span: Span,
     ) -> SourceResult<Self> {
         // Number of content columns: Always at least one.
         let c = tracks.x.len().max(1);
 
-        // If not all columns in the last row have cells, we will add empty
-        // cells and complete the row so that those positions are susceptible
-        // to show rules and receive grid styling.
-        // We apply '% c' twice so that 'cells_remaining' is zero when
-        // the last row is already filled (then 'cell_count % c' would be zero).
+        // We can't just use the cell's index in the 'cells' vector to
+        // determine its automatic position, since cells could have arbitrary
+        // positions, so the position of a cell in 'cells' can differ from its
+        // final position in 'resolved_cells' (see below).
+        // Therefore, we use a counter, 'auto_index', to determine the position
+        // of the next cell with (x: auto, y: auto). It is only stepped when
+        // a cell with (x: auto, y: auto), usually the vast majority, is found.
+        let mut auto_index = 0;
+
+        // We have to rebuild the grid to account for arbitrary positions.
+        // Create at least 'cells.len()' positions, since there will be at
+        // least 'cells.len()' cells, even though some of them might be placed
+        // in arbitrary positions and thus cause the grid to expand.
         let cell_count = cells.len();
-        let cells_remaining = (c - cell_count % c) % c;
-        let cells = cells
-            .iter()
-            .cloned()
-            .chain(std::iter::repeat_with(T::default).take(cells_remaining))
+        let mut resolved_cells: Vec<Option<Cell>> = Vec::with_capacity(cell_count);
+        for cell in cells.iter().cloned() {
+            // Let's calculate the cell's final position based on its
+            // requested position.
+            let resolved_index = {
+                let cell_x = cell.x(styles);
+                let cell_y = cell.y(styles);
+                resolve_cell_position(cell_x, cell_y, &resolved_cells, &mut auto_index, c)
+                    .at(span)?
```

Comment:
This should probably also use the cell's span.

If cell spans turn out to be detached, we might need to adjust the place where they are constructed and add an appropriate .spanned().

Reply:
I've implemented an initial version of this. For now I just made it use the table's span when the cell's span is detached. How do you suggest I go about this?

> If cell spans turn out to be detached, we might need to adjust the place where they are constructed and add an appropriate .spanned().

How could this be achieved? Do you mean going through all places in the standard library which build a `GridCell` or `TableCell` and adding `.spanned()` accordingly? Like e.g. in bibliography?

Reply:
Okay, I changed bibliography.rs, as far as I could search this is the only place where we create `GridCell` internally, so I made them use the bibliography element's span.